subsys/sd/mmc.c

```diff
@@ -343,72 +343,88 @@ static inline int mmc_set_max_freq(struct sd_card *card, struct sd_csd *card_csd
 	return 0;
 }
 
-static int mmc_set_timing(struct sd_card *card)
+static int mmc_set_bus_width(struct sd_card *card)
 {
 	int ret;
 	struct sdhc_command cmd = {0};
 
-	/* Change Card Timing Mode */
+	if (card->host_props.host_caps.bus_8_bit_support) {
+		cmd.arg = MMC_SWITCH_8_BIT_BUS_ARG;
+		card->bus_io.bus_width = SDHC_BUS_WIDTH8BIT;
+	} else if (card->host_props.host_caps.bus_4_bit_support) {
+		cmd.arg = MMC_SWITCH_4_BIT_BUS_ARG;
+		card->bus_io.bus_width = SDHC_BUS_WIDTH4BIT;
+	} else {
+		/* If only 1 bit bus is supported, nothing to be done */
+		return 0;
+	}
+
+	/* Set Card Bus Width */
 	cmd.opcode = SD_SWITCH;
-	cmd.arg = MMC_SWITCH_HS_TIMING_ARG;
 	cmd.response_type = SD_RSP_TYPE_R1b;
 	cmd.timeout_ms = CONFIG_SD_CMD_TIMEOUT;
 	ret = sdhc_request(card->sdhc, &cmd, NULL);
+	sdmmc_wait_ready(card);
 	if (ret) {
-		LOG_DBG("Error setting bus timing: %d", ret);
+		LOG_ERR("Setting card data bus width failed: %d", ret);
 		return ret;
 	}
-	sdmmc_wait_ready(card);
 
-	/* Max frequency in HS mode is 52 MHz */
-	card->bus_io.clock = 52000000;
-	card->bus_io.timing = SDHC_TIMING_HS;
-	/* Change SDHC bus timing */
+	/* Set host controller bus width */
 	ret = sdhc_set_io(card->sdhc, &card->bus_io);
 	if (ret) {
+		LOG_ERR("Setting SDHC data bus width failed: %d", ret);
 		return ret;
 	}
 
-	return 0;
+	return ret;
 }
 
-static int mmc_set_bus_width(struct sd_card *card)
+static int mmc_set_HS_timing(struct sd_card *card)
```

Comment:
minor naming nit: I'd rather see this named `mmc_set_hs_timing` or `mmc_set_high_speed_timing`